### PR TITLE
Allow skipping symlink patterns, to avoid raising a fault

### PR DESCRIPTION
**Pull request by sanmai-NL**
## Description

In #14627, Prettier was redesigned to ‘no longer follow symbolic links’. However, the implementation went further and turned into a breaking change. Rather than adhering to the [exit status contract](https://prettier.io/docs/en/cli#exit-codes) of only returning a non-zero exit status when a file is not formatted or when something is wrong with Prettier itself, #14627 made Prettier return a non-zero exit status upon any pattern turning out to be a pointing to symlink.

This PR fixes the original PR by @andersk to behave in line with its documentation.

## Checklist

- [x] I’ve added tests to confirm my change works.
- [x] (If changing the API or CLI) I’ve documented the changes I’ve made (in the `docs/` directory).
- [x] (If the change is user-facing) I’ve added my changes to `changelog_unreleased/*/XXXX.md` file following `changelog_unreleased/TEMPLATE.md`.
- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

**Comment by sanmai-NL** [comment]
@fisker Can you approve to kick off CI?

**Comment by fisker** [comment]
I think the behavior is expected, but maybe we can skip it if `--no-error-on-unmatched-pattern` is used.

**Comment by sanmai-NL** [comment]
@fisker, Can you comment on the fact that the behavior, does not match the docs and was different up to the breaking change?

**Comment by sanmai-NL** [comment]
@fisker This a reminder. My team is waiting for this.

**Comment by fisker** [comment]
If I  remember  correctly, it only throws the the pattern is used as arguments.

Say you have `a/b` is a symblink. The error only throws when you use `prettier a/b` or `prettier a/b someother-pattern`. It doesn't throws if you use `prettier a` or `prettier "a/**/*".

Correct me if I'm wrong.

**Comment by sanmai-NL** [comment]
This was clarified in https://github.com/prettier/prettier/pull/14627#issuecomment-1679118689.

A tool that uses e.g., Git index changes to trigger Prettier cannot filter for symlinks in a cross-OS compatible manner.

**Comment by fisker** [comment]
Put it behind `--no-error-on-unmatched-pattern` then.  https://github.com/prettier/prettier/pull/15533#issuecomment-1792409044

**Comment by sanmai-NL** [comment]
What do you mean with ‘it’? Do you mean changing this PR to fix `--no-error-on-unmatched-pattern` or just using `--no-error-on-unmatched-pattern` in the first place?

This invocation throws a fault, and we have been using that parameter this all along. This is also discussed in the PR, that you wanted me to open a new issue for.

```text
│ [lefthook] executing: prettier --list-different --no-error-on-unmatched-pattern --ignore-unknown -- .lefthook.yml .mega-linter.yml .sourcery.yaml .v8rrc.yml
┃  prettier ❯ 

[error] Explicitly specified pattern ".lefthook.yml" is a symbolic link.
[error] Explicitly specified pattern ".mega-linter.yml" is a symbolic link.
[error] Explicitly specified pattern ".sourcery.yaml" is a symbolic link.
[error] Explicitly specified pattern ".v8rrc.yml" is a symbolic link.
```

**Comment by fisker** [comment]
> I think the behavior is expected, but maybe we can skip it if `--no-error-on-unmatched-pattern` is used.

NOT THROW IF `--no-error-on-unmatched-pattern`.

**Comment by sosukesuzuki** [comment]
@fisker Please allow me to ask a question for clarification. Is what you are meaning following?

It is expected behavior that Prettier throws an error when you directly specify the symbolic link path.
However, we can accept pull requests that change that it does not throw errors when `--no-error-on-unmatched-pattern` is enabled.

**Comment by fisker** [comment]
Yes, that's what I expect, what do you think?

**Comment by sosukesuzuki** [comment]
I'm :+1: on your opinion.

File: changelog_unreleased/cli/15533.md
Comment (andersk) [inline changelog_unreleased/cli/15533.md:3]: I take exception to this mischaracterization of #14627. Prettier has always exited with code 2 for an unsupported file (e.g. `touch foo.xyz; prettier foo.xyz`). With #14627, symlinks are intentionally unsupported files, so it is appropriate to exit with code 2. The implementation adheres to the contract, and this behavior was specifically [noted and agreed upon](https://github.com/prettier/prettier/pull/14627#pullrequestreview-1364278508) with the maintainer.

You could easily describe what you’re changing here without implying that I made some kind of mistake.

File: src/cli/expand-patterns.js
Comment (andersk) [inline src/cli/expand-patterns.js:74]: The request was to avoid throwing an error if `--no-error-on-unmatched-pattern` is set, not to continue processing the file as if it were regular.

File: changelog_unreleased/cli/15533.md
Comment (sanmai-NL) [reply changelog_unreleased/cli/15533.md:3]: It's a pity you only bring forward your argument now the PR is finished, even as you were pinged in the OP. In this PR or indeed in reply to my comments on your earlier PR, you could've responded with factual discussion.

An unmatched file is simply a branch of the business logic, an has nothing to do with a fault state. If what you're claiming is correct, the docs aren't correct in not taking the unmatched files fault (parameter default) into account.

Comment (fisker) [reply changelog_unreleased/cli/15533.md:3]: Stop arguing about what's correct or wrong. I'll take PR this over.

Comment (andersk) [reply changelog_unreleased/cli/15533.md:3]: I [did bring forward this argument](https://github.com/prettier/prettier/pull/14627#issuecomment-1754003937). It was clear even before my comments that your PR isn’t finished, as it’s failing CI; in any case, receiving and responding to review feedback is a normal, healthy part of the PR process. If you think some part of my discussion is not factual, you are welcome to explain.

I’ll repeat that #14627 *did not introduce* the convention that runtime errors exit with code 2. It seems to me that you’re reading too much into the documentation’s [very few words](https://prettier.io/docs/en/cli#exit-codes) on exit codes that it doesn’t actually say. If we consider the (again, pre-existing) case of the specified file being missing, for whatever argument you might have that it’s not precisely described by some narrow interpretation of code 1 “Something wasn’t formatted properly” or code 2 “Something’s wrong with Prettier”, there’s a much clearer case that it’s not correctly described by code 0 “Everything formatted properly”. But if you’d like to improve the documentation, my suggestion would be to open a separate PR for that (I’ll defer to the maintainers if they have a different suggestion).

My main concern in this comment is that I believe it’s unfair to characterize #14627 in the way that you have.

Comment (sanmai-NL) [reply changelog_unreleased/cli/15533.md:3]: @andersk , you make a circular argument. You argue against my PR changing behavior from your PR, just because it conflicts with your PR (‘With https://github.com/prettier/prettier/pull/14627, symlinks are intentionally unsupported files’). That behavior change was breaking and disables a valid use-case, as I already argued in your PR thread. I also discussed and you conceded that a proper enhancement would be to process symlinks and separate invalid/unacceptable ones from acceptable ones.

> The implementation adheres to the contract (which never said “Prettier itself”—

I quote the docs about exit status 2:

> Something’s wrong with Prettier

Period. My description ‘Prettier _itself_’ adds no meaning,  it only stresses the exact meaning of this sentence. 

> this is one of many runtime errors that are completely reasonable to categorize there)

Okay, so what do you think about this scale:

1. Invalid memory access.
2. Division by zero.
3. Non-existent configuration parameter
4. All files specified are unmatched.
5. One file unmatched because it does not exist.
6. Some files are deemed unmatched because they're broken or unacceptable symlinks.
7. One file among matched ones, is unmatched just because it is a symlink.

You made anything from 6 on result in a fault (an error is something a human does). For case 4. extended to, at most, unsupported files, one could stretch the meaning of fault, since there would be no files to check left. Just as well we could let Prettier log that and exit normally. But 7 is by no means a fault. Especially not when it is a perfectly valid symlink.

Comment (sanmai-NL) [reply changelog_unreleased/cli/15533.md:3]: @fisker  I didn't see your comment before responding. I will rest my case, even though I think there are more claims left here by @andersk that I disagree with.

File: changelog_unreleased/cli/15533.md
Comment (sanmai-NL) [inline changelog_unreleased/cli/15533.md:3]: Typo: ... be filter_ed_ out ..

Comment (sanmai-NL) [reply changelog_unreleased/cli/15533.md:3]: to prevent Prettier to throw errors.

Comment (fisker) [reply changelog_unreleased/cli/15533.md]: Thanks.

**Review by fisker — APPROVED** [review]

**Review by andersk — APPROVED** [review]

**Comment by sanmai-NL** [comment]
I'm fine with them too.

**Comment by sanmai-NL** [comment]
Should @sosukesuzuki still review this? Hopefully users like in #15723 can be helped out soon. The PR is very small and @fisker made some fixes and improvements already.

**Comment by junaga** [comment]
1. resolve all symlinks
2. add an option for no symlink resolving

you are not going to patch every existing tool in the world into prettier lmao.

**Review by sosukesuzuki — APPROVED** [review]
Sorry for delay my review, LGTM

**Pull request merged**
